core/src/main/java/org/apache/iceberg/rest/auth/OAuth2Util.java

```diff
@@ -550,31 +512,29 @@ public static AuthSession fromAccessToken(
         RESTClient client,
         ScheduledExecutorService executor,
         String token,
-        Long defaultExpiresInMillis,
+        Long defaultExpiresAtMillis,
         AuthSession parent) {
       AuthSession session =
           new AuthSession(
               parent.headers(), token, OAuth2Properties.ACCESS_TOKEN_TYPE, parent.credential());
 
-      long startTimeMillis = System.currentTimeMillis();
-      Long expiresAtMillis = session.expiresAtMillis();
+      Long expiresAtMillis = null;
 
-      if (isExpired(expiresAtMillis, startTimeMillis)) {
+      if (session.expiresAtMillis() != null
+          && session.expiresAtMillis() <= System.currentTimeMillis()) {
+        long startTimeMillis = System.currentTimeMillis();
         Pair<Integer, TimeUnit> expiration = session.refresh(client);
         if (expiration != null) {
-          scheduleTokenRefresh(
-              client, executor, session, startTimeMillis, expiration.first(), expiration.second());
+          expiresAtMillis = startTimeMillis + expiration.second().toMillis(expiration.first());
```

Comment:
Now that token refresh is based around `expiresAt`, should we prefer the specific timestamp from the token if it exists? I think it would make sense to check `session.expiresAtMillis()` here and use it in place of the duration, since the duration requires careful handling of start time.

Reply:
that makes sense

Reply:
I've updated the code and noticed that we probably still have to check both: the expiration time from the refreshed token (if it's set and if token refresh was successful), and then fall back to the expiration that we get from the response.